File: src/main/java/com/keygenqt/mylibrary/api/validators/LoginValidator.java
package com.keygenqt.mylibrary.api.validators;

import com.keygenqt.mylibrary.api.bodies.LoginBody;
import com.keygenqt.mylibrary.base.CustomValidate;
import com.keygenqt.mylibrary.models.repositories.UserRepository;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Component;
import org.springframework.validation.Errors;
import org.springframework.validation.Validator;

@Component
public class LoginValidator implements Validator {

    private final UserRepository repository;

    @Autowired
    public LoginValidator(UserRepository repository) {
        this.repository = repository;
    }

    @Override
    public boolean supports(@NotNull Class<?> clazz) {
        return LoginBody.class == clazz;
    }

    @Override
    public void validate(@NotNull Object target, @NotNull Errors errors) {

        CustomValidate.isRequired("email", target, errors);
        CustomValidate.isEmail("email", target, errors);
        CustomValidate.isPassword("password", target, errors);
        CustomValidate.isRequired("uid", target, errors);

        if (!errors.hasErrors()) {
            var body = (LoginBody) target;
            var user = repository.findAllByEmail(body.getEmail());
            if (user == null) {
                CustomValidate.setError("email", CustomValidate.MESSAGE_FOUND_EMPTY, errors);
            } else {
                if (!new BCryptPasswordEncoder().matches(body.getPassword(), user.getPassword())) {
                    CustomValidate.setError("password", CustomValidate.MESSAGE_FIELD_INCORRECT, errors);
                }
            }
        }
    }
}

File: src/main/kotlin/com/keygenqt/mylibrary/api/validators/LoginValidator.kt
package com.keygenqt.mylibrary.api.validators

import com.keygenqt.mylibrary.base.*
import com.keygenqt.mylibrary.models.repositories.*
import org.springframework.beans.factory.annotation.*
import org.springframework.security.crypto.bcrypt.*
import org.springframework.stereotype.*
import org.springframework.validation.*

@Component
class LoginValidator : Validator {

    @Autowired
    private lateinit var repository: UserRepository

    override fun supports(clazz: Class<*>): Boolean {
        return Login::class.java == clazz
    }

    override fun validate(target: Any, errors: Errors) {

        errors.validateEmail(target)
        errors.validatePassword(target)
        errors.validateRequired(target, "uid")

        if (!errors.hasErrors() && target is Login) {
            repository.findAllByEmail(target.email ?: "")?.let { model ->
                if (!BCryptPasswordEncoder().matches(target.password, model.password)) {
                    "field.incorrect".let { errors.rejectValue("password", it, BaseMessageUtils.getMessage(it)) }
                }
            } ?: run {
                "field.found.empty".let { errors.rejectValue("email", it, BaseMessageUtils.getMessage(it)) }
            }
        }
    }
}